django/core/management/commands/shell.py

```diff
@@ -111,10 +115,35 @@ def python(self, options):
         # Start the interactive interpreter.
         code.interact(local=imported_objects)
 
+    def get_and_report_namespace(self, verbosity):
+        namespace = self.get_namespace()
+
+        if verbosity >= 1:
+            self.stdout.write(
+                f"{len(namespace)} objects imported automatically",
+                self.style.SUCCESS,
+            )
+
+        return namespace
+
+    def get_namespace(self):
+        apps_models = apps.get_models()
+        apps_models_modules = [
+            (app.models_module, app.label) for app in apps.get_app_configs()
+        ]
+        namespace = {}
+        for model in reversed(apps_models):
+            if model.__module__:
+                namespace[model.__name__] = model
+        for app in apps_models_modules:
+            app_models_module, app_label = app
+            namespace[f"{app_label}_models"] = app_models_module
+        return namespace
+
     def handle(self, **options):
         # Execute the command and exit.
         if options["command"]:
-            exec(options["command"], globals())
+            exec(options["command"], {**self.get_namespace()})
```

Comment:
```suggestion
            exec(options["command"], {{**globals(), **self.get_namespace()})
```
We are losing the globals (same below)

Previously we were testing that `'print("__name__" in globals())'` returned `True` - this is no longer the case

I think you can update the tests to
```diff
--- a/tests/shell/tests.py
+++ b/tests/shell/tests.py
@@ -23,7 +23,7 @@ from . import models as shell_models
 
 
 class ShellCommandTestCase(SimpleTestCase):
-    script_globals_import = 'print("Phone" in globals())'
+    script_globals = 'print("__name__" in globals() and "Phone" in globals())'
     script_with_inline_function = (
         "import django\ndef f():\n    print(django.__version__)\nf()"
     )
@@ -39,16 +39,16 @@ class ShellCommandTestCase(SimpleTestCase):
             )
         self.assertEqual(cm.records[0].getMessage(), __version__)
 
+    def test_command_option_globals(self):
+        with captured_stdout() as stdout:
+            call_command("shell", command=self.script_globals)
+        self.assertEqual(stdout.getvalue().strip(), "True")
+
     def test_command_option_inline_function_call(self):
         with captured_stdout() as stdout:
             call_command("shell", command=self.script_with_inline_function)
         self.assertEqual(stdout.getvalue().strip(), __version__)
 
-    def test_command_option_with_imports(self):
-        with captured_stdout() as stdout:
-            call_command("shell", command=self.script_globals_import)
-        self.assertEqual(stdout.getvalue().strip(), "True")
-
     @unittest.skipIf(
         sys.platform == "win32", "Windows select() doesn't support file descriptors."
     )
@@ -65,9 +65,9 @@ class ShellCommandTestCase(SimpleTestCase):
         "Windows select() doesn't support file descriptors.",
     )
     @mock.patch("django.core.management.commands.shell.select")  # [1]
-    def test_stdin_read_globals_import(self, select):
+    def test_stdin_read_globals(self, select):
         with captured_stdin() as stdin, captured_stdout() as stdout:
-            stdin.write(self.script_globals_import)
+            stdin.write(self.script_globals)
             stdin.seek(0)
             call_command("shell")
         self.assertEqual(stdout.getvalue().strip(), "True")
```
(a lot of this is to reduce the diff for reviewing)

Reply:
We delete globals on purpose following this [discussion](https://github.com/django/django/pull/18158#discussion_r1636925531)

Reply:
https://github.com/django/django/pull/13660#discussion_r520365761
This appears to be the original discussion on why it was added
It's certainly safer to keep it and I would feel more comfortable keeping it 🤔

Reply:
Ok, if it is safer to keep it I'll keep it 👍 .